kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/types/DataTypeParser.java

```diff
@@ -0,0 +1,234 @@
+    }
+
+    /**
+     * Parses an <a href="https://github.com/delta-io/delta/blob/master/PROTOCOL.md#struct-type">
+     *     struct type </a>
+     */
+    private static StructType parseStructType(JsonNode json) {
+        checkArgument(json.isObject() && json.size() == 2,
+            String.format(
+                "Expected JSON object with 2 fields for struct data type but got:\n%s", json));
+        JsonNode fieldsNode = getNonNullField(json, "fields");
+        checkArgument(fieldsNode.isArray(),
+            String.format("Expected array for fieldName=%s in:\n%s", "fields", json));
+        Iterator<JsonNode> fields = fieldsNode.elements();
+        List<StructField> parsedFields = new ArrayList<>();
+        while (fields.hasNext()) {
+            parsedFields.add(parseStructField(fields.next()));
+        }
+        return new StructType(parsedFields);
+    }
+
+    /**
+     * Parses an <a href="https://github.com/delta-io/delta/blob/master/PROTOCOL.md#struct-field">
+     *     struct field </a>
+     */
+    private static StructField parseStructField(JsonNode json) {
+        checkArgument(json.isObject(), "Expected JSON object for struct field");
+        String name = getStringField(json, "name");
+        DataType type = parseDataType(getNonNullField(json, "type"));
+        boolean nullable = getBooleanField(json, "nullable");
+        Map<String, String> metadata = parseFieldMetadata(json.get("metadata"));
+        return new StructField(
+            name,
+            type,
+            nullable,
+            metadata
+        );
+    }
+
+    // TODO for now we maintain the current behavior that parses field metadata as a
+    //  Map<String, String> for either string or numerical value types. A follow-up PR will refactor
+    //  this to support all the supported value types for field metadata and add a FieldMetadata
+    //  class in place of Map<String, String>
+    private static Map<String, String> parseFieldMetadata(JsonNode json) {
+        if (json == null || json.isNull()) {
+            return Collections.emptyMap();
+        }
+
+        checkArgument(json.isObject(), "Expected JSON object for struct field metadata");
+        final Iterator<Map.Entry<String,JsonNode>> iterator = json.fields();
+        final Map<String, String> metadata = new HashMap<>();
+        while (iterator.hasNext()) {
+            Map.Entry<String, JsonNode> entry = iterator.next();
+            JsonNode value = entry.getValue();
+            String key = entry.getKey();
+
+            if (!(value.isTextual() || value.isIntegralNumber())) {
+                throw new UnsupportedOperationException(
+                    "Only numerical or string type field metadata values are supported");
+            }
+            metadata.put(key, value.asText());
+        }
+        return metadata;
+    }
+
+    private static String FIXED_DECIMAL = "decimal\\(\\s*(\\d+)\\s*,\\s*(\\-?\\d+)\\s*\\)";
```

Comment:
what should _really_ be static here isn't just the string, but the `Pattern.compile(FIXED_DECIMAL)`. That's the expensive part that we don't want to recompute every `nameToType` invocation

Reply:
Fixed